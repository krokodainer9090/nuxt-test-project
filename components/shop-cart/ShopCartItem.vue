<script>
// Components
import Input from '../controls/input/TheInput';

export default {
  name: 'ShopCartItem',

  components: {
    Input
  },

  props: {
    cartItem: {
      type: Object,
      default: _ => ({})
    }
  },

  data: () => ({
    isError: false
  }),

  computed: {
    currentItemCount: {
      get () {
        return String(this.cartItem.currentCount);
      },

      set (currentCount) {
        if ( currentCount <= this.cartItem.count ) {
          this.isError = false;

          this.$emit('increase-count', {
            id: this.cartItem.id,
            count: currentCount
          });
        } else {
          this.isError = true;
        }
      }
    },

    itemName () {
      return this.cartItem?.N || '';
    },

    itemPrice () {
      return this.cartItem?.price
        ? `${this.cartItem.price} руб.`
        : '';
    }
  },

  methods: {
    deleteItem () {
      this.$emit('delete-item', this.cartItem.id);
    }
  }
};
</script>

<template>
  <div class="shop-cart-item">
    <div class="row shop-cart-item__content">
      <div class="col-6">
        <span
          v-if="itemName"
          class="shop-cart-item__name"
        >
          {{ itemName }}
        </span>
      </div>
      <div class="col-2">
        <Input
          v-model="currentItemCount"
          type="number"
          class="shop-cart-item__input"
        />
        <div
          v-if="isError"
          class="shop-cart-item__error"
        >
          <span class="shop-cart-item__error-text">
            Столько товара на складе нет. Осталось товара: {{ cartItem.count }} шт.
          </span>
        </div>
      </div>
      <div class="col-2 col-md-3">
        <span
          v-if="itemPrice"
          class="shop-cart-item__price"
        >
          {{ itemPrice }}
          <span class="shop-cart-item__price-count">/ шт.</span>
        </span>
      </div>
      <div class="col-2 col-md-1">
        <div
          class="shop-cart-item__delete-btn"
          @click="deleteItem"
        >
          <span class="shop-cart-item__delete-btn-text">
            Удалить
          </span>
        </div>
      </div>
    </div>
  </div>
</template>

<style lang="scss">
@import "../../styles/components/shop-cart/shop-cart-item";
</style>
