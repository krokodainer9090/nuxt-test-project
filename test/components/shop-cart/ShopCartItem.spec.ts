// Components
import ShopCartItem from '~/components/shop-cart/ShopCartItem.vue';

// Utilities
import { shallowMount, Wrapper } from '@vue/test-utils';
import { shopCatalogItem } from '~/test/utils/misc/constants';

// Constants
const CLASS_NAME = '.shop-cart-item';

const classes = {
  input: `${CLASS_NAME}__input`,
  deleteBtn: `${CLASS_NAME}__delete-btn`
};

describe('ShopCartItem', () => {
  const mountFunction: (options?: Object) => Wrapper<any> = (options = {}) => {
    return shallowMount(ShopCartItem, {
      ...options
    });
  };

  it('should render empty item', () => {
    const wrapper = mountFunction();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render item', async () => {
    const wrapper = mountFunction({
      propsData: {
        cartItem: shopCatalogItem
      }
    });
    const emitObject = wrapper.emitted();

    expect(wrapper.html()).toMatchSnapshot();

    // Check increase count of item
    await wrapper.setData({
      currentItemCount: shopCatalogItem.count
    });

    expect(emitObject['increase-count']?.[0]).toEqual([{
      id: shopCatalogItem.id,
      count: shopCatalogItem.count
    }]);

    // Check delete item emit
    await wrapper.find(classes.deleteBtn).trigger('click');

    expect(emitObject['delete-item']?.[0]).toEqual([shopCatalogItem.id]);
  });

  it('should handler error currentCount', async () => {
    const wrapper = mountFunction({
      propsData: {
        cartItem: shopCatalogItem
      }
    });
    const emitObject = wrapper.emitted();

    // Check error increase count of item
    await wrapper.setData({
      currentItemCount: shopCatalogItem.count + 1
    });

    // // No emit if error
    expect(emitObject['increase-count']?.[0]).toBeUndefined();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
