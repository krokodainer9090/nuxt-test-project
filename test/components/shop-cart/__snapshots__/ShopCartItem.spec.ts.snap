// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopCartItem should handler error currentCount 1`] = `
"<div class=\\"shop-cart-item\\">
  <div class=\\"row shop-cart-item__content\\">
    <div class=\\"col-6\\"><span class=\\"shop-cart-item__name\\">
        Суп с пекинской капустой
      </span></div>
    <div class=\\"col-2\\">
      <input-stub id=\\"1\\" value=\\"0\\" type=\\"number\\" min=\\"0\\" class=\\"shop-cart-item__input\\"></input-stub>
      <div class=\\"shop-cart-item__error\\"><span class=\\"shop-cart-item__error-text\\">
          Столько товара на складе нет. Осталось товара: 15 шт.
        </span></div>
    </div>
    <div class=\\"col-2 col-md-3\\"><span class=\\"shop-cart-item__price\\">
        42.8 руб.
        <span class=\\"shop-cart-item__price-count\\">/ шт.</span></span></div>
    <div class=\\"col-2 col-md-1\\">
      <div class=\\"shop-cart-item__delete-btn\\"><span class=\\"shop-cart-item__delete-btn-text\\">
          Удалить
        </span></div>
    </div>
  </div>
</div>"
`;

exports[`ShopCartItem should render empty item 1`] = `
"<div class=\\"shop-cart-item\\">
  <div class=\\"row shop-cart-item__content\\">
    <div class=\\"col-6\\">
      <!---->
    </div>
    <div class=\\"col-2\\">
      <input-stub id=\\"1\\" value=\\"undefined\\" type=\\"number\\" min=\\"0\\" class=\\"shop-cart-item__input\\"></input-stub>
      <!---->
    </div>
    <div class=\\"col-2 col-md-3\\">
      <!---->
    </div>
    <div class=\\"col-2 col-md-1\\">
      <div class=\\"shop-cart-item__delete-btn\\"><span class=\\"shop-cart-item__delete-btn-text\\">
          Удалить
        </span></div>
    </div>
  </div>
</div>"
`;

exports[`ShopCartItem should render item 1`] = `
"<div class=\\"shop-cart-item\\">
  <div class=\\"row shop-cart-item__content\\">
    <div class=\\"col-6\\"><span class=\\"shop-cart-item__name\\">
        Суп с пекинской капустой
      </span></div>
    <div class=\\"col-2\\">
      <input-stub id=\\"1\\" value=\\"0\\" type=\\"number\\" min=\\"0\\" class=\\"shop-cart-item__input\\"></input-stub>
      <!---->
    </div>
    <div class=\\"col-2 col-md-3\\"><span class=\\"shop-cart-item__price\\">
        42.8 руб.
        <span class=\\"shop-cart-item__price-count\\">/ шт.</span></span></div>
    <div class=\\"col-2 col-md-1\\">
      <div class=\\"shop-cart-item__delete-btn\\"><span class=\\"shop-cart-item__delete-btn-text\\">
          Удалить
        </span></div>
    </div>
  </div>
</div>"
`;
